### PR TITLE
Update forest tutorial with mano, venv and forest updates

### PR DESCRIPTION
**Pull request by GeorgeEfstathiadis**

### CODE REVIEW

File: tutorials/helper_functions.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    if time_end is None:
+        time_end = datetime.today().strftime("%Y-%m-%d") + "T23:59:00"
+
+    if users == []:
+        print("Obtaining list of users...")
+        num_tries = 1
+        while num_tries < 5:
+            try:
+                users = [u for u in mano.users(keyring, study_id)]
+                num_tries = 6
```
Comment (hackdna) [inline tutorials/helper_functions.py:177]: Replace with:
```
else:
    break
```

File: tutorials/helper_functions.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            except KeyboardInterrupt:
+                print("Someone closed the program")
+                sys.exit()
+            except mano.APIError as e:
+                print("Something is wrong with your credentials:")
+                print(e)
+                num_tries = 6
```
Comment (hackdna) [inline tutorials/helper_functions.py:184]: `break`

File: tutorials/helper_functions.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+    if not os.path.isdir(download_folder):
+        os.mkdir(download_folder)
+
+    if time_end is None:
+        time_end = datetime.today().strftime("%Y-%m-%d") + "T23:59:00"
+
+    if users == []:
+        print("Obtaining list of users...")
+        num_tries = 1
```
Comment (hackdna) [inline tutorials/helper_functions.py:173]: `num_tries` = 0`?

File: tutorials/helper_functions.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                print(e)
+                num_tries = 6
+            except Exception:
+                num_tries = num_tries + 1
+
+    for u in users:
+        zf = None
+        download_success = False
+        num_tries = 0
+        while not download_success:
```
Comment (hackdna) [inline tutorials/helper_functions.py:192]: `while num_tries < 5:`?

File: tutorials/helper_functions.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+                    keyring,
+                    study_id,
+                    u,
+                    data_streams,
+                    time_start=time_start,
+                    time_end=time_end,
+                )
+                if zf is not None:
+                    zf.extractall(download_folder)
+                download_success = True
```
Comment (hackdna) [inline tutorials/helper_functions.py:205]: `break`?

File: tutorials/helper_functions.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                num_tries = num_tries + 1
+            except KeyboardInterrupt:
+                print("Someone closed the program")
+                sys.exit()
+            except mano.APIError as e:
+                print("Something is wrong with your credentials:")
+                print(e)
+                download_success = True
+            if num_tries > 5:
+                download_success = True
```
Comment (hackdna) [inline tutorials/helper_functions.py:217]: `break`?

File: tutorials/helper_functions.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            except requests.exceptions.ChunkedEncodingError:
+                print(f"Network failed in download of {u}, try {num_tries}")
+                num_tries = num_tries + 1
+            except KeyboardInterrupt:
+                print("Someone closed the program")
+                sys.exit()
+            except mano.APIError as e:
+                print("Something is wrong with your credentials:")
+                print(e)
+                download_success = True
```
Comment (hackdna) [inline tutorials/helper_functions.py:215]: `break`?

File: tutorials/keyring_studies.py
Comment (hackdna) [inline tutorials/keyring_studies.py:1]: Since this only sets shell env vars, should that just be a shell script?

**Review by hackdna — CHANGES_REQUESTED** [review]

File: tutorials/keyring_studies.py
Comment (GeorgeEfstathiadis) [reply tutorials/keyring_studies.py:1]: So I added the shell script that replaces `keyring_studies.py`, but I am unsure how to run it from jupyter notebook and have it access the environment variables from the notebook. I tried simply `!beiwe_credentials.sh` and I have set up its access to be executable by everyone, but yet I don't seem to be able to access the environment variables that way.

Comment (hackdna) [reply tutorials/keyring_studies.py:1]: I see. Running a shell script in a notebook happens in a subprocess, so changes to env vars won't be reflected in the notebook itself. I guess in this case it's fine to have a Python script, or may be use `%env`?
https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-env

File: tutorials/helper_functions.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        keyring["USERNAME"] == ""
+        or keyring["PASSWORD"] == ""
+        or keyring["ACCESS_KEY"] == ""
+        or keyring["SECRET_KEY"] == ""
+    ):
+        print("Error: Did you set up the keyring_studies.py file?")
+        return
+
+    if not os.path.isdir(download_folder):
+        os.makedirs(download_folder, exist_ok=True)
```
Comment (hackdna) [inline tutorials/helper_functions.py:167]: The point of `exist_ok=True` is that you don't need to check if a directory exists.
https://docs.python.org/3.8/library/os.html#os.makedirs

File: tutorials/helper_functions.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    if time_end is None:
+        time_end = datetime.today().strftime("%Y-%m-%d") + "T23:59:00"
+
+    if users is None:
+        print("Obtaining list of users...")
+        num_tries = 0
+        while num_tries < 5:
+            try:
+                users = [u for u in mano.users(keyring, study_id)]
+                break
```
Comment (hackdna) [inline tutorials/helper_functions.py:178]:
```
else:
    break
```
https://docs.python.org/3/reference/compound_stmts.html#except

File: tutorials/helper_functions.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            try:
+                zf = mano.sync.download(
+                    keyring,
+                    study_id,
+                    u,
+                    data_streams,
+                    time_start=time_start,
+                    time_end=time_end,
+                )
+                break
```
Comment (hackdna) [inline tutorials/helper_functions.py:203]:
```
else:
    break
```
https://docs.python.org/3/reference/compound_stmts.html#except

File: tutorials/helper_functions.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+                num_tries += 1
+            except KeyboardInterrupt:
+                print("Someone closed the program")
+                sys.exit()
+            except mano.APIError as e:
+                print("Something is wrong with your credentials:")
+                print(e)
+                break
+
+        if num_tries == 5:
```
Comment (hackdna) [inline tutorials/helper_functions.py:215]: I think while-else clause would be more readable
https://realpython.com/lessons/while-loop-else-clause/

**Review by hackdna — CHANGES_REQUESTED** [review]
Thanks!

File: tutorials/keyring_studies.py
Comment (GeorgeEfstathiadis) [reply tutorials/keyring_studies.py:1]: I am not sure how to use the `%env` magic to set environment variables that are hidden in the script.

Comment (GeorgeEfstathiadis) [reply tutorials/keyring_studies.py:1]: because we want the actual variable values to be set in a separate place to share between scripts

**Review by hackdna — APPROVED** [review]
Thank you!

**Pull request merged**